File: cc/app/acc-serv/internal/proto/s5/tcp.cc
//
// weproxy@foxmail.com 2022/10/03
//

#include "fx/fx.h"
#include "gx/net/net.h"
#include "gx/time/time.h"
#include "logx/logx.h"
#include "nx/netio/netio.h"
#include "nx/socks/socks.h"
#include "nx/stats/stats.h"
#include "s5.h"

namespace internal {
namespace proto {
namespace s5 {
using namespace nx;

////////////////////////////////////////////////////////////////////////////////
// handleTCP ...
void handleTCP(net::Conn c, net::Addr raddr) {
    auto tag = GX_SS(TAG << " TCP_" << nx::NewID() << " " << c->RemoteAddr() << "->" << raddr);
    auto sta = stats::NewTCPStats(stats::Type::S5, tag);

    sta->Start("connected");
    DEFER(sta->Done("closed"));

    // dial target server
    AUTO_R(rc, er1, net::Dial(raddr));
    if (er1) {
        LOGS_E(TAG << " dial, err: " << er1);
        socks::WriteReply(c, socks::Reply::HostUnreachable);
        return;
    }
    DEFER(rc->Close());

    // <<< REP:
    //     | VER | CMD |  RSV  | ATYP | BND.ADDR | BND.PORT |
    //     +-----+-----+-------+------+----------+----------+
    //     |  1  |  1  | X'00' |  1   |    ...   |    2     |
    auto err = socks::WriteReply(c, socks::Reply::Success, 0, net::MakeAddr(net::IPv4zero, 0));
    if (err) {
        if (err != net::ErrClosed) {
            LOGS_E(TAG << " err: " << err);
        }
        return;
    }

    netio::RelayOption opt(time::Second * 2);
    opt.A2B.CopingFn = [sta](int n) { sta->AddRecv(n); };
    opt.B2A.CopingFn = [sta](int n) { sta->AddSent(n); };

    // Relay c <--> rc
    netio::Relay(c, rc, opt);
}

////////////////////////////////////////////////////////////////////////////////
// handleUDP ...
extern void handleUDP(net::PacketConn c, net::Addr caddr, net::Addr raddr);

// handleAssoc ...
void handleAssoc(net::Conn c, net::Addr raddr) {
    auto caddr = c->RemoteAddr();

    auto tag = GX_SS(TAG << " Assoc_" << nx::NewID() << " " << caddr << "->" << raddr);
    auto sta = stats::NewTCPStats(stats::Type::S5, tag);

    sta->Start("connected");
    DEFER(sta->Done("closed"));

    AUTO_R(ln, er1, net::ListenPacket(":0"));
    if (er1) {
        LOGS_E(TAG << " dial, err: " << er1);
        socks::WriteReply(c, socks::Reply::HostUnreachable);
        return;
    }

    // handleUDP
    gx::go([ln, caddr, raddr] {
        // raddr is not fixed, it maybe be changed after
        handleUDP(ln, caddr, raddr);
    });

    // <<< REP:
    //     | VER | CMD |  RSV  | ATYP | BND.ADDR | BND.PORT |
    //     +-----+-----+-------+------+----------+----------+
    //     |  1  |  1  | X'00' |  1   |    ...   |    2     |
    auto err = socks::WriteReply(c, socks::Reply::Success, 0, ln->LocalAddr());
    if (err) {
        if (err != net::ErrClosed) {
            LOGS_E(TAG << " err: " << err);
        }
        return;
    }

    // discard
    io::Copy(io::Discard, c);
}

}  // namespace s5
}  // namespace proto
}  // namespace internal

File: go/app/acc-serv/internal/proto/s5/tcp.go
//
// weproxy@foxmail.com 2022/10/20
//

package s5

import (
	"errors"
	"fmt"
	"io"
	"net"

	"weproxy/acc/libgo/logx"
	"weproxy/acc/libgo/nx"
	"weproxy/acc/libgo/nx/netio"
	"weproxy/acc/libgo/nx/socks"
	"weproxy/acc/libgo/nx/stats"
)

////////////////////////////////////////////////////////////////////////////////

// handleTCP ...
func handleTCP(c net.Conn, raddr net.Addr) {
	tag := fmt.Sprintf("%s TCP_%v %v.%v", TAG, nx.NewID(), c.RemoteAddr(), raddr)
	sta := stats.NewTCPStats(stats.TypeS5, tag)

	sta.Start("connected")
	defer sta.Done("closed")

	// dial target server
	rc, err := net.Dial("tcp", raddr.String())
	if err != nil {
		logx.E("%s dial, err: %v", TAG, err)
		socks.WriteReply(c, socks.ReplyHostUnreachable, 0, nil)
		return
	}
	rc.(*net.TCPConn).SetKeepAlive(true)
	defer rc.Close()

	// <<< REP:
	//     | VER | CMD |  RSV  | ATYP | BND.ADDR | BND.PORT |
	//     +-----+-----+-------+------+----------+----------+
	//     |  1  |  1  | X'00' |  1   |    ...   |    2     |
	err = socks.WriteReply(c, socks.ReplySuccess, 0, &net.TCPAddr{IP: net.IPv4zero, Port: 0})
	if err != nil {
		if !errors.Is(err, net.ErrClosed) {
			logx.E("%s err: ", TAG, err)
		}
		return
	}

	opt := netio.RelayOption{}
	opt.A2B.CopingFn = func(n int) { sta.AddRecv(int64(n)) }
	opt.B2A.CopingFn = func(n int) { sta.AddSent(int64(n)) }

	// Relay c <--> rc
	netio.Relay(c, rc, opt)
}

////////////////////////////////////////////////////////////////////////////////

// handleAssoc ...
func handleAssoc(c net.Conn, raddr net.Addr) {
	caddr := c.RemoteAddr()

	tag := fmt.Sprintf("%s Assoc_%v %v.%v", TAG, nx.NewID(), caddr, raddr)
	sta := stats.NewTCPStats(stats.TypeS5, tag)

	sta.Start("connected")
	defer sta.Done("closed")

	ln, err := net.ListenPacket("udp", ":0")
	if err != nil {
		logx.E("%s dial err: %v", tag, err)
		socks.WriteReply(c, socks.ReplyHostUnreachable, 0, nil)
		return
	}

	// handleUDP
	go handleUDP(ln)

	// <<< REP:
	//     | VER | CMD |  RSV  | ATYP | BND.ADDR | BND.PORT |
	//     +-----+-----+-------+------+----------+----------+
	//     |  1  |  1  | X'00' |  1   |    ...   |    2     |
	err = socks.WriteReply(c, socks.ReplySuccess, 0, ln.LocalAddr())
	if err != nil {
		if !errors.Is(err, net.ErrClosed) {
			logx.E("%s err: %v", tag, err)
		}
		return
	}

	// discard
	io.Copy(io.Discard, c)
}
